File: src/plnflr/importing/roomplan.py
"""RoomPlan USDZ → floor polygon, object holes, door thresholds."""

from __future__ import annotations

import io
import math
import zipfile
from collections import defaultdict
from dataclasses import dataclass

from plnflr.domain.models import Ring, Room, Threshold, Vertex
from plnflr.domain.units import metres_to_mm_coord
from plnflr.engine.clip import normalize_ring
from plnflr.importing.usda import UsdaMesh, parse_usda_mesh, transform_point

THRESHOLD_MM = 80
_TOP_Z = 0.02
_ROOM_NAMES = {
    "livingroom": "Salon",
    "bedroom": "Sypialnia",
    "bathroom": "Łazienka",
    "kitchen": "Kuchnia",
    "diningroom": "Jadalnia",
}


@dataclass(frozen=True, slots=True)
class CapturedScan:
    name: str
    room: Room
    thresholds: tuple[Threshold, ...]
    vertices_m: str
    hole_rectangles: str
    door_vertices: str


def room_from_usdz(payload: bytes) -> CapturedScan:
    try:
        archive = zipfile.ZipFile(io.BytesIO(payload))
    except zipfile.BadZipFile as exc:
        raise ValueError("to nie jest RoomPlan USDZ") from exc
    names = archive.namelist()
    if "Scan.usda" not in names:
        raise ValueError("USDZ nie jest skanem RoomPlan (brak Scan.usda)")
    floors = [name for name in names if "/Floors/Floor" in name and name.endswith(".usda")]
    if not floors:
        raise ValueError("RoomPlan: brak podłogi")
    floor = parse_usda_mesh(archive.read(floors[0]).decode(), name="Floor")
    outline_m = _floor_outline_m(floor)
    origin = (
        min(point[0] for point in outline_m),
        min(point[1] for point in outline_m),
    )
    outer = _ring_mm(outline_m, origin)
    fireplaces: list[Ring] = []
    fireplace_rects: list[str] = []
    thresholds: list[Threshold] = []
    door_polys: list[str] = []
    for name in names:
        if not name.endswith(".usda") or "/Mesh/" not in name:
            continue
        text = archive.read(name).decode()
        mesh = parse_usda_mesh(text, name=name.rsplit("/", 1)[-1].removesuffix(".usda"))
        category = mesh.category.lower()
        if category.startswith("fireplace"):
            ring, rect = _box_ring(mesh, origin)
            fireplaces.append(ring)
            fireplace_rects.append(rect)
        elif category.startswith("door"):
            threshold = _door_threshold(mesh, origin)
            thresholds.append(threshold)
            door_polys.append(_ring_to_vertices_m(threshold.geometry))
    holes = tuple(fireplaces) + tuple(item.geometry for item in thresholds)
    return CapturedScan(
        name=_room_name(archive.read("Scan.usda").decode()),
        room=Room(outer, holes=holes),
        thresholds=tuple(thresholds),
        vertices_m=_ring_to_vertices_m(outer),
        hole_rectangles="\n".join(fireplace_rects),
        door_vertices="\n\n".join(door_polys),
    )


def _room_name(root: str) -> str:
    for key, label in _ROOM_NAMES.items():
        if f"{key}0" in root.lower():
            return label
    return "Pokój"


def _floor_outline_m(floor: UsdaMesh) -> list[tuple[float, float]]:
    top = [index for index, point in enumerate(floor.points) if abs(point[2]) <= _TOP_Z]
    top_set = set(top)
    if len(top) < 3:
        raise ValueError("RoomPlan: podłoga nie ma płaszczyzny")
    cursor = 0
    edge_count: dict[frozenset[int], int] = defaultdict(int)
    for count in floor.counts:
        face = floor.indices[cursor : cursor + count]
        cursor += count
        if count != 3 or not all(index in top_set for index in face):
            continue
        for start, end in ((face[0], face[1]), (face[1], face[2]), (face[2], face[0])):
            edge_count[frozenset((start, end))] += 1
    boundary: dict[int, list[int]] = defaultdict(list)
    for edge, seen in edge_count.items():
        if seen != 1:
            continue
        a, b = tuple(edge)
        boundary[a].append(b)
        boundary[b].append(a)
    if not boundary:
        raise ValueError("RoomPlan: nie da się odczytać obrysu podłogi")
    start = min(boundary, key=lambda index: (_plan(floor, index), index))
    walked = [start]
    previous = start
    current = boundary[start][0]
    while current != start:
        walked.append(current)
        nxt = next(node for node in boundary[current] if node != previous)
        previous, current = current, nxt
        if len(walked) > len(floor.points):
            raise ValueError("RoomPlan: obrys podłogi jest uszkodzony")
    return [_plan(floor, index) for index in walked]


def _plan(mesh: UsdaMesh, index: int) -> tuple[float, float]:
    world = transform_point(mesh.points[index], mesh.matrix)
    return world[0], world[2]


def _box_corners_m(mesh: UsdaMesh) -> list[tuple[float, float]]:
    seen: set[tuple[float, float]] = set()
    corners: list[tuple[float, float]] = []
    for point in mesh.points:
        world = transform_point((point[0], 0.0, point[2]), mesh.matrix)
        key = (round(world[0], 4), round(world[2], 4))
        if key in seen:
            continue
        seen.add(key)
        corners.append((world[0], world[2]))
    if len(corners) < 3:
        raise ValueError(f"RoomPlan: {mesh.name} nie ma rzutu na podłogę")
    return corners


def _ordered_ring_m(points: list[tuple[float, float]]) -> list[tuple[float, float]]:
    cx = sum(point[0] for point in points) / len(points)
    cy = sum(point[1] for point in points) / len(points)
    return sorted(points, key=lambda point: math.atan2(point[1] - cy, point[0] - cx))


def _box_ring(mesh: UsdaMesh, origin: tuple[float, float]) -> tuple[Ring, str]:
    ring = _ring_mm(_ordered_ring_m(_box_corners_m(mesh)), origin)
    return ring, _ring_to_rect_m(ring)


def _door_threshold(mesh: UsdaMesh, origin: tuple[float, float]) -> Threshold:
    corners = _ordered_ring_m(_box_corners_m(mesh))
    ring = _ring_mm(corners, origin)
    length_mm, width_mm, long_i = _rect_axes(ring)
    if width_mm < THRESHOLD_MM:
        ring = _expand_width(ring, long_i, THRESHOLD_MM)
        width_mm = THRESHOLD_MM
    return Threshold(
        label="listwa progowa",
        length_mm=length_mm,
        width_mm=width_mm,
        geometry=ring,
    )


def _rect_axes(ring: Ring) -> tuple[int, int, int]:
    verts = ring.vertices
    edges = []
    for index, vertex in enumerate(verts):
        nxt = verts[(index + 1) % len(verts)]
        length = int(round(math.hypot(nxt.x_mm - vertex.x_mm, nxt.y_mm - vertex.y_mm)))
        edges.append((length, index))
    long_len, long_i = max(edges)
    short_len = min(edge[0] for edge in edges)
    return long_len, short_len, long_i


def _expand_width(ring: Ring, long_i: int, width_mm: int) -> Ring:
    verts = ring.vertices
    a = verts[long_i]
    b = verts[(long_i + 1) % len(verts)]
    dx, dy = b.x_mm - a.x_mm, b.y_mm - a.y_mm
    length = math.hypot(dx, dy) or 1.0
    nx, ny = -dy / length, dx / length
    _, current, _ = _rect_axes(ring)
    extra = (width_mm - current) / 2
    grown = []
    cx = sum(v.x_mm for v in verts) / len(verts)
    cy = sum(v.y_mm for v in verts) / len(verts)
    for vertex in verts:
        side = 1 if (vertex.x_mm - cx) * nx + (vertex.y_mm - cy) * ny >= 0 else -1
        grown.append(
            Vertex(
                int(round(vertex.x_mm + side * extra * nx)),
                int(round(vertex.y_mm + side * extra * ny)),
            )
        )
    return Ring(tuple(grown))


def _ring_mm(points: list[tuple[float, float]], origin: tuple[float, float]) -> Ring:
    verts = tuple(
        Vertex(
            metres_to_mm_coord(f"{point[0] - origin[0]:.6f}"),
            metres_to_mm_coord(f"{point[1] - origin[1]:.6f}"),
        )
        for point in points
    )
    path = normalize_ring(tuple((v.x_mm, v.y_mm) for v in verts))
    start = min(range(len(path)), key=lambda i: (path[i][0], path[i][1], i))
    rotated = path[start:] + path[:start]
    return Ring(tuple(Vertex(x, y) for x, y in rotated))


def _ring_to_vertices_m(ring: Ring) -> str:
    return "\n".join(f"{v.x_mm / 1000:.3f},{v.y_mm / 1000:.3f}" for v in ring.vertices)


def _ring_to_rect_m(ring: Ring) -> str:
    xs = [v.x_mm for v in ring.vertices]
    ys = [v.y_mm for v in ring.vertices]
    min_x, min_y = min(xs), min(ys)
    width_m = (max(xs) - min_x) / 1000
    height_m = (max(ys) - min_y) / 1000
    return f"{min_x / 1000:.3f},{min_y / 1000:.3f},{width_m:.3f},{height_m:.3f}"

File: Kernel/Sources/PlnFlrLayout/RoomPlan.swift
import Foundation

public struct CapturedRoom: Codable, Equatable, Sendable {
    public var name: String
    public var room: Room
    public var thresholds: [Threshold]
    public var windows: [Opening]

    public init(name: String, room: Room, thresholds: [Threshold], windows: [Opening]) {
        self.name = name
        self.room = room
        self.thresholds = thresholds
        self.windows = windows
    }
}

public struct CapturedScan: Equatable, Sendable {
    public var rooms: [CapturedRoom]

    public init(rooms: [CapturedRoom]) {
        self.rooms = rooms
    }

    public var name: String { rooms.first?.name ?? "Pokój" }
    public var room: Room { rooms[0].room }
    public var thresholds: [Threshold] { rooms.flatMap(\.thresholds) }
    public var windows: [Opening] { rooms.flatMap(\.windows) }
    public var verticesM: String {
        rooms.first.map { ringToVerticesM($0.room.outer) } ?? ""
    }
    public var holeRectangles: String {
        guard let first = rooms.first else { return "" }
        let doors = Set(first.thresholds.map { ringToVerticesM($0.geometry) })
        return first.room.holes
            .filter { !doors.contains(ringToVerticesM($0)) }
            .map(ringToRectM)
            .joined(separator: "\n")
    }
    public var doorVertices: String {
        rooms.first?.thresholds.map { ringToVerticesM($0.geometry) }.joined(separator: "\n\n") ?? ""
    }
    public var windowSegments: String {
        rooms.first?.windows.map(openingToSegmentM).joined(separator: "\n") ?? ""
    }
}

private let thresholdMm = 80
private let topZ = 0.02
private let roomNames: [String: String] = [
    "livingroom": "Salon",
    "bedroom": "Sypialnia",
    "bathroom": "Łazienka",
    "kitchen": "Kuchnia",
    "diningroom": "Jadalnia",
]

public func roomFromUsdz(_ payload: Data) throws -> CapturedScan {
    let archive: ZipArchive
    do {
        archive = try ZipArchive(data: payload)
    } catch {
        throw ScanError.notZip
    }
    let names = archive.names
    guard names.contains("Scan.usda") || names.contains("./Scan.usda") else {
        throw ScanError.notRoomPlan
    }
    let floors = names.filter { $0.contains("/Floors/Floor") && $0.hasSuffix(".usda") }.sorted()
    guard !floors.isEmpty else { throw ScanError.missingFloor }
    var outlinesM: [[(Double, Double)]] = []
    for floorName in floors {
        guard let floorData = archive[floorName], let floorText = String(data: floorData, encoding: .utf8)
        else { throw ScanError.missingFloor }
        let mesh = try parseUsdaMesh(floorText, name: floorName)
        outlinesM.append(try floorOutlineM(mesh))
    }
    let origin = (
        outlinesM.flatMap { $0.map(\.0) }.min() ?? 0,
        outlinesM.flatMap { $0.map(\.1) }.min() ?? 0
    )
    let floorsMm = try outlinesM.map { try ringMm($0, origin: origin) }
    var fireplaces: [Ring] = []
    var thresholds: [Threshold] = []
    var windows: [Opening] = []
    for name in names {
        guard name.hasSuffix(".usda"), name.contains("/Mesh/") else { continue }
        if name.contains("/Floors/Floor") { continue }
        guard let bytes = archive[name], let text = String(data: bytes, encoding: .utf8) else { continue }
        let leaf = name.split(separator: "/").last.map { String($0.replacingOccurrences(of: ".usda", with: "")) } ?? name
        let mesh = try parseUsdaMesh(text, name: leaf)
        let category = mesh.category.lowercased()
        if category.hasPrefix("fireplace") {
            fireplaces.append(try boxRing(mesh, origin: origin).0)
        } else if category.hasPrefix("door") {
            thresholds.append(try doorThreshold(mesh, origin: origin))
        } else if category.hasPrefix("window") {
            windows.append(try windowOpening(mesh, origin: origin))
        }
    }
    let scanText = String(data: archive["Scan.usda"] ?? archive["./Scan.usda"] ?? Data(), encoding: .utf8) ?? ""
    let labels = roomLabels(scanText, count: floorsMm.count)
    var rooms: [CapturedRoom] = []
    for (index, floor) in floorsMm.enumerated() {
        let roomFire = fireplaces.filter { featureTouches($0, floor) }
        let roomDoors = thresholds.filter { featureTouches($0.geometry, floor) }
        let roomWindows = windows.filter { windowTouches($0, floor) }
        rooms.append(
            CapturedRoom(
                name: labels[index],
                room: Room(floor, holes: roomFire + roomDoors.map(\.geometry)),
                thresholds: roomDoors,
                windows: roomWindows
            )
        )
    }
    return CapturedScan(rooms: rooms)
}

private func roomLabels(_ root: String, count: Int) -> [String] {
    let pattern = try! NSRegularExpression(pattern: #"def Xform \"([^\"]+)\""#)
    let ns = root as NSString
    let order = ["livingroom", "bedroom", "bathroom", "kitchen", "diningroom"]
    var found: [String] = []
    for match in pattern.matches(in: root, range: NSRange(location: 0, length: ns.length)) {
        let raw = ns.substring(with: match.range(at: 1)).lowercased()
        for key in order where raw.contains(key) {
            found.append(roomNames[key]!)
            break
        }
    }
    if found.isEmpty {
        return (0..<count).map { $0 == 0 ? "Pokój" : "Pokój \($0 + 1)" }
    }
    while found.count < count {
        found.append("Pokój \(found.count + 1)")
    }
    return Array(found.prefix(count))
}

private func featureTouches(_ feature: Ring, _ floor: Ring) -> Bool {
    let box = bbox(Room(floor))
    let xs = feature.vertices.map(\.xMm)
    let ys = feature.vertices.map(\.yMm)
    let minX = xs.min() ?? 0, maxX = xs.max() ?? 0
    let minY = ys.min() ?? 0, maxY = ys.max() ?? 0
    return maxX >= box.minX && minX <= box.maxX && maxY >= box.minY && minY <= box.maxY
}

private func windowTouches(_ opening: Opening, _ floor: Ring) -> Bool {
    let box = bbox(Room(floor))
    let minX = min(opening.start.xMm, opening.end.xMm)
    let maxX = max(opening.start.xMm, opening.end.xMm)
    let minY = min(opening.start.yMm, opening.end.yMm)
    let maxY = max(opening.start.yMm, opening.end.yMm)
    return maxX >= box.minX && minX <= box.maxX && maxY >= box.minY && minY <= box.maxY
}

private func floorOutlineM(_ floor: UsdaMesh) throws -> [(Double, Double)] {
    let top = floor.points.indices.filter { abs(floor.points[$0].2) <= topZ }
    let topSet = Set(top)
    guard top.count >= 3 else { throw ScanError.noFloorPlane }
    var cursor = 0
    var edgeCount: [Set<Int>: Int] = [:]
    for count in floor.counts {
        let face = Array(floor.indices[cursor..<(cursor + count)])
        cursor += count
        guard count == 3, face.allSatisfy({ topSet.contains($0) }) else { continue }
        for pair in [(face[0], face[1]), (face[1], face[2]), (face[2], face[0])] {
            let key: Set<Int> = [pair.0, pair.1]
            edgeCount[key, default: 0] += 1
        }
    }
    var boundary: [Int: [Int]] = [:]
    for (edge, seen) in edgeCount where seen == 1 {
        let nodes = Array(edge)
        guard nodes.count == 2 else { continue }
        boundary[nodes[0], default: []].append(nodes[1])
        boundary[nodes[1], default: []].append(nodes[0])
    }
    guard !boundary.isEmpty else { throw ScanError.noOutline }
    let start = boundary.keys.min { a, b in
        let pa = plan(floor, a), pb = plan(floor, b)
        if pa.0 != pb.0 { return pa.0 < pb.0 }
        if pa.1 != pb.1 { return pa.1 < pb.1 }
        return a < b
    }!
    var walked = [start]
    var previous = start
    var current = boundary[start]![0]
    while current != start {
        walked.append(current)
        guard let nxt = boundary[current]?.first(where: { $0 != previous }) else {
            throw ScanError.brokenOutline
        }
        previous = current
        current = nxt
        if walked.count > floor.points.count { throw ScanError.brokenOutline }
    }
    return walked.map { plan(floor, $0) }
}

private func plan(_ mesh: UsdaMesh, _ index: Int) -> (Double, Double) {
    let world = transformPoint(mesh.points[index], matrix: mesh.matrix)
    return (world.0, world.2)
}

private func boxCornersM(_ mesh: UsdaMesh) throws -> [(Double, Double)] {
    var seen = Set<String>()
    var corners: [(Double, Double)] = []
    for point in mesh.points {
        let world = transformPoint((point.0, 0.0, point.2), matrix: mesh.matrix)
        let key = String(format: "%.4f,%.4f", world.0, world.2)
        if seen.contains(key) { continue }
        seen.insert(key)
        corners.append((world.0, world.2))
    }
    if corners.count < 3 { throw ScanError.noPlan(mesh.name) }
    return corners
}

private func orderedRingM(_ points: [(Double, Double)]) -> [(Double, Double)] {
    let cx = points.map(\.0).reduce(0, +) / Double(points.count)
    let cy = points.map(\.1).reduce(0, +) / Double(points.count)
    return points.sorted { atan2($0.1 - cy, $0.0 - cx) < atan2($1.1 - cy, $1.0 - cx) }
}

private func boxRing(_ mesh: UsdaMesh, origin: (Double, Double)) throws -> (Ring, String) {
    let ring = try ringMm(orderedRingM(try boxCornersM(mesh)), origin: origin)
    return (ring, ringToRectM(ring))
}

private func windowOpening(_ mesh: UsdaMesh, origin: (Double, Double)) throws -> Opening {
    let corners = try orderedRingM(try boxCornersM(mesh)).map {
        (max(0.0, $0.0 - origin.0), max(0.0, $0.1 - origin.1))
    }
    let ring = try ringMm(corners, origin: (0, 0))
    let verts = ring.vertices
    var edges: [(Double, Int, Int)] = []
    for index in verts.indices {
        let nxt = verts[(index + 1) % verts.count]
        let length = pow(Double(nxt.xMm - verts[index].xMm), 2) + pow(Double(nxt.yMm - verts[index].yMm), 2)
        let wall = min(verts[index].xMm + verts[index].yMm, nxt.xMm + nxt.yMm)
        edges.append((length, -wall, index))
    }
    let longI = edges.max { a, b in
        if a.0 != b.0 { return a.0 < b.0 }
        if a.1 != b.1 { return a.1 < b.1 }
        return a.2 < b.2
    }!.2
    return Opening(label: "okno", start: verts[longI], end: verts[(longI + 1) % verts.count])
}

private func openingToSegmentM(_ opening: Opening) -> String {
    String(
        format: "%.3f,%.3f,%.3f,%.3f",
        Double(opening.start.xMm) / 1000,
        Double(opening.start.yMm) / 1000,
        Double(opening.end.xMm) / 1000,
        Double(opening.end.yMm) / 1000
    )
}

private func doorThreshold(_ mesh: UsdaMesh, origin: (Double, Double)) throws -> Threshold {
    var ring = try ringMm(orderedRingM(try boxCornersM(mesh)), origin: origin)
    var (lengthMm, widthMm, longI) = rectAxes(ring)
    if widthMm < thresholdMm {
        ring = expandWidth(ring, longI: longI, widthMm: thresholdMm)
        widthMm = thresholdMm
    }
    return Threshold(label: "listwa progowa", lengthMm: lengthMm, widthMm: widthMm, geometry: ring)
}

private func rectAxes(_ ring: Ring) -> (Int, Int, Int) {
    let verts = ring.vertices
    var edges: [(Int, Int)] = []
    for index in verts.indices {
        let nxt = verts[(index + 1) % verts.count]
        let dx = Double(nxt.xMm - verts[index].xMm)
        let dy = Double(nxt.yMm - verts[index].yMm)
        edges.append((Int((dx * dx + dy * dy).squareRoot().rounded()), index))
    }
    let long = edges.max { $0.0 < $1.0 }!
    let short = edges.map(\.0).min()!
    return (long.0, short, long.1)
}

private func expandWidth(_ ring: Ring, longI: Int, widthMm: Int) -> Ring {
    let verts = ring.vertices
    let a = verts[longI]
    let b = verts[(longI + 1) % verts.count]
    let dx = Double(b.xMm - a.xMm)
    let dy = Double(b.yMm - a.yMm)
    let length = (dx * dx + dy * dy).squareRoot() == 0 ? 1.0 : (dx * dx + dy * dy).squareRoot()
    let nx = -dy / length
    let ny = dx / length
    let current = rectAxes(ring).1
    let extra = Double(widthMm - current) / 2
    let cx = Double(verts.map(\.xMm).reduce(0, +)) / Double(verts.count)
    let cy = Double(verts.map(\.yMm).reduce(0, +)) / Double(verts.count)
    let grown = verts.map { vertex -> Vertex in
        let side = (Double(vertex.xMm) - cx) * nx + (Double(vertex.yMm) - cy) * ny >= 0 ? 1.0 : -1.0
        return Vertex(
            Int((Double(vertex.xMm) + side * extra * nx).rounded()),
            Int((Double(vertex.yMm) + side * extra * ny).rounded())
        )
    }
    return try! Ring(grown)
}

private func ringMm(_ points: [(Double, Double)], origin: (Double, Double)) throws -> Ring {
    let verts = try points.map { point in
        Vertex(
            try metresToMmCoord(String(format: "%.6f", point.0 - origin.0)),
            try metresToMmCoord(String(format: "%.6f", point.1 - origin.1))
        )
    }
    var path = normalizeRing(verts)
    let start = path.indices.min { a, b in
        if path[a].xMm != path[b].xMm { return path[a].xMm < path[b].xMm }
        if path[a].yMm != path[b].yMm { return path[a].yMm < path[b].yMm }
        return a < b
    } ?? 0
    path = Array(path[start...]) + Array(path[..<start])
    return try Ring(path)
}

private func ringToVerticesM(_ ring: Ring) -> String {
    ring.vertices.map { String(format: "%.3f,%.3f", Double($0.xMm) / 1000, Double($0.yMm) / 1000) }
        .joined(separator: "\n")
}

private func ringToRectM(_ ring: Ring) -> String {
    let xs = ring.vertices.map(\.xMm)
    let ys = ring.vertices.map(\.yMm)
    let minX = xs.min() ?? 0
    let minY = ys.min() ?? 0
    let widthM = Double((xs.max() ?? 0) - minX) / 1000
    let heightM = Double((ys.max() ?? 0) - minY) / 1000
    return String(format: "%.3f,%.3f,%.3f,%.3f", Double(minX) / 1000, Double(minY) / 1000, widthM, heightM)
}
